tools/assemble_sequence.cpp

```diff
@@ -0,0 +1,1334 @@
+#include <filesystem>
```

Comment:
I have doubts on a lot of the dedicated tools introduced to build the audio data ELF files, could you elaborate on why they are needed? The GNU assembler should be adequately equipped to do the same thing, as far as I can tell, and is much less dependency-wise (as we already use it)

Reply:
I'm not clear what you mean by using the assembler.  Do you mean for building the audio tables, or for building the actual audio data files themselves?

Reply:
To be specific, I mean audio sequences and the tables that were previously in the `data` directory. I don't know enough about the structure of the audiobank and audiotable to know exactly how applicable it is there, but I suspect it should be at least somewhat applicable there for creating ELF files from the audio data output by the vadpcm encoder?

Reply:
Via the assembler or via objcopy?  You could theoretically do it all through the assembler, yes, but it would be really cumbersome to do.  If we used an assembler process, then you would still need to use the Python scripts/CPP program to generate the raw binary representations of the assets (no way to get the assembler to do that except with sequences, and the user experience there is iffy at best), and then you'd need an .s file to define the start/end for every single file and inject the binary contents, and then separately handle that in a file.  It would _work_ for sure, it's just much more manual and less modding-friendly than having scripts that handle it all for you.

Reply:
Could you describe why the user experience is not good w.r.t. assembling sequences with GAS? Back when a sequence decoder was first made I had a go at matching audioseq with it and I think it turned out rather well (https://github.com/thar0/oot/commit/4916993c2ee9485b51f690e25d80c31d4535315b)
The only thing that stands out is the sequence table, which when I tried I didn't try to get it working cleanly, but I can't imagine that it would be any more involved than the process for modifying anything in `include/tables/`, which is so far the standard.

I can't describe in any detail currently what a process might look like for audiobank and audiotable since as I mentioned I need to get up to speed with their structure, but for audio samples surely
`aiff -> aifc -> incbin -> as -> ld`
all handled by the makefile isn't a whole lot different to python scripts to the end user? I think it's preferable to leverage the makefile as much as possible rather than offloading a substantial portion of the build system onto scripts that may not be as understandable and necessitate additional dependencies. Further, if our build system is going to support alternative endians and pointer widths as the scripts appear to have such options, GNU binutils has such functionality built-in, requiring no maintenance from us.

Reply:
Just to add for reference, here's what the structure looks like for soundFonts (audiobank), blue boxes being relative addresses/offsets from the start of the soundfonts, and red boxes being relative addresses/offsets for the samples (audiotable). I think there's also a blue-box highlight missing in drums (the 0x230 in [0x230, 0.31498f], but I seem to have lost this sheet since I put it together).

![image](https://user-images.githubusercontent.com/47598039/169738246-3e1940f4-b29c-4961-93f1-7f858424a09b.png)

Reply:
The scripts can provide better contextual error messages if someone is making their own assets, rather than relying on a context-free assembler.  That was one of the original motivations (coming from conversations that happened months ago here 😅).  Also, there's no intuitive way to convert an AIFC into assembly without manually handling a good portion of it, since it isn't as straightforward as taking binary contents from the AIFC and concatenating it to the sample banks.  That's not even looking at soundfonts; there's pretty much no way to convert an XML into the appropriate binary representation via GAS, and if you use Python to do so, it already needs to open the AIFC files because the soundfonts contain the metadata for the samples inside of them.

So basically, you would still need the scripts, but we could potentially strip out the ELF part and do the final packaging with GAS.  You lose the endian handling, since GAS won't handle that with .incbin as far as I'm aware (???), and you could do the audio tables in assembly if you want, but it doesn't seem to gain you anything if the scripts already have all that context.

Reply:
FYI I replaced the ELF creation stuff in #1509 by emitting C code / tables.